src/tfs/tlog.c

```diff
@@ -1,180 +1,418 @@
+
+    /* sector offset, length and staging buffer of current extension */
+    range sectors;
+    buffer staging;
+    buffer tuple_staging;
+    u64 tuple_bytes_remain;
+    boolean extension_open;
+
     int dirty;              /* cas boolean */
     heap h;
 } *log;
 
-closure_function(1, 1, void, log_write_completion,
-                 vector, v,
+closure_function(3, 1, void, log_write_completion,
+                 buffer, b,
+                 vector, completions,
+                 boolean, release,
                  status, s)
 {
     // reclaim the buffer now and the vector...make it a whole thing
-    status_handler i;
-    int len = vector_length(bound(v));
-    for (int count = 0; count < len; count++) {
-        i = vector_delete(bound(v), 0);
-        apply(i, s);
-    }
+    vector v = bound(completions);
+    status_handler sh;
+
+    vector_foreach(v, sh)
+        apply(sh, s);
+
+    deallocate_vector(v);
+
+    if (bound(release))
+        deallocate_buffer(bound(b));
     closure_finish();
 }
 
 // xxx  currently we cant take writes during the flush
 
-/* XXX it's not right to just stick SECTOR_{SIZE,OFFSET} everywhere...
-   and add block_log2 to fs */
-range log_block_range(log tl, u64 length)
+/* Avoid references to log, which may be in transition to a new extension. */
+static void log_flush_internal(heap h, filesystem fs, buffer b, range log_range,
+                               vector completions, boolean release)
 {
-    return irange(tl->offset >> SECTOR_OFFSET,
-                  (tl->offset + pad(length, tl->fs->blocksize)) >> SECTOR_OFFSET);
+    push_u8(b, END_OF_LOG);
+
+#ifdef TLOG_DEBUG_DUMP
+    u64 z = b->end;
+    b->start = 0;
+    b->end = 1024;
+    rprintf("staging contains:\n%X\n", b);
+    b->end = z;
+#endif
+
+    assert(buffer_length(b) > 0); /* END_OF_LOG, at least */
+    assert((b->start & MASK(fs->blocksize_order)) == 0);
+    u64 sector_start = log_range.start + sector_from_offset(fs, b->start);
+    u64 sectors = sector_from_offset(fs, buffer_length(b) + (fs_blocksize(fs) - 1));
+    range write_range = irange(sector_start, sector_start + sectors);
+    assert(range_contains(log_range, write_range));
+
+    void *p = buffer_ref(b, 0);
+    tlog_debug("log_flush_internal: writing sectors %R, buffer addr %p\n", write_range, p);
+    apply(fs->w, p, write_range, closure(h, log_write_completion, b, completions, release));
+    if (!release) {
+        b->end -= 1;                /* next write removes END_OF_LOG */
+        tlog_debug("log ext offset was %d (end %d)\n", b->start, b->end);
+        b->start = b->end & ~MASK(fs->blocksize_order); /* next storage write starting here */
+        tlog_debug("log ext offset now %d\n", b->start);
+        assert(b->end >= b->start);
+    }
 }
 
-/* XXX we're just writing the whole log - instead write only from
-   block of buffer start forward */
 void log_flush(log tl)
 {
-    if (!__sync_bool_compare_and_swap(&tl->dirty, 1, 0))
+    if (!tl->dirty)
```

Comment:
Not sure where that line was from...but yeah, if anything this would eventually be under a 'tfs' medium-grain lock, so I don't see the point in anything fine-grained here.